### PR TITLE
Add support for tagging certificates. Fix deprecated tasks in aws_acm integration tests

### PR DESCRIPTION
**Pull request by sebastien-rosset**
##### SUMMARY

This PR adds support for configuring arbitrary tags when importing a certificate using the `aws_acm` module. Previously, it was only possible to set the 'Name' tag.

Additionally, this PR fixes issues with the `aws_acm` integration tests.  The integration tests were using deprecated tasks or attributes, such as `openssl_certificate`.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

aws_acm

##### ADDITIONAL INFORMATION

Changes to the `aws_acm.py` module:
1. Add new `tags` and `purge_tags` attributes.
2. The `certificate_arn` attribute is now allowed when `state='present'`. A playbook should be allowed to modify an existing certificate entry by providing the ARN. For example, a play may want to add, modify, remove tags on an existing certificate.
3. The `aws_acm` module returns the updated tags. See example below.
4. Refactor `aws_acm.py` to improve code reuse and make it possible to set arbitrary tags. This should also help to 1) improve readability. 2) prepare for #869 which I am planning to work on next.

Backwards-compatibility is retained, even though it might make sense to normalize some of the attributes.

Example return value:
```
"certificate": {
            "arn": "arn:aws:acm:us-west-1:account:certificate/f85abf9d-4bda-4dcc-98c3-770664a68243",
            "domain_name": "acm1.949058644.ansible.com",
            "tags": {
                "Application": "search",
                "Environment": "development",
                "Name": "ansible-test-78006277-398b5796f999_949058644_1"
            }
        }
```

**Integration tests:**
1. The `openssl_certificate` task is deprecated. Migrate to `x509_certificate`.
2. The `signature_algorithms` attribute is no longer supported by the new `x509_certificate` task. Using `selfsigned_digest` instead.
3. The integration tests for the `aws_acm` module pass locally.
4. I see https://github.com/ansible/ansible/issues/67788 has been closed, but tests/integration/targets/aws_acm/aliases still has `unstable`. I am not sure what to do about it. I was able to run the tests in my local workspace after making the above changes.

```paste below

```

### CODE REVIEW

File: plugins/modules/aws_acm.py
Comment (markuman) [inline plugins/modules/aws_acm.py:165]: ...same `version_added` for `purge_tags`

**Review by markuman — CHANGES_REQUESTED** [review]
Good work! Can you also a a changelog fragment?

**Comment by sebastien-rosset** [comment]
There may be an intermittent timing bug in [plugins/module_utils/acm.py:get_certificates](https://github.com/ansible-collections/amazon.aws/blob/main/plugins/module_utils/acm.py#L93) of the `amazon.aws` repo.

The `get_certificate` function retrieves a list of certificates using `list_certificates_with_backoff`. For each cert ARN returned by `list_certificates_with_backoff`, the function invokes `describe_certificate_with_backoff`. It's possible that between the call to `list_certificates_with_backoff` and `describe_certificate_with_backoff`, one of the certificates has been deleted. In that case, a `ResourceNotFoundException` exception is raised. The exception does not seem to be caught. Therefore any ansible task that invokes the `get_certificates` function fails.

I'm still trying to understand the behavior of this decorator. I thought it meant the `ResourceNotFoundException` exception would be caught, but apparently it's not.
```python
@AWSRetry.backoff(tries=5, delay=5, backoff=2.0, catch_extra_error_codes=['ResourceNotFoundException', 'RequestInProgressException'])
```

To reproduce, the following tasks can be executed multiple times until it fails:


```yaml
  - name: delete cert
    aws_acm:
      state: absent
      domain_name: 'mydoamain.com'
  - name: check cert was deleted
    aws_acm_info:
      tags:
        Name: 'my_certificate_name'
```

This happens in the integration tests here: https://github.com/ansible-collections/community.aws/blob/main/tests/integration/targets/aws_acm/tasks/main.yml#L262-L267

Here is a stack trace when the exception occurs:
```
TASK [aws_acm : check cert 3 was deleted] ********************************************************************************************************************
task path: /root/ansible_collections/community/aws/tests/output/.tmp/integration/aws_acm-uxj6vjlf-ÅÑŚÌβŁÈ/tests/integration/targets/aws_acm/tasks/main.yml:435
Using module file /root/ansible_collections/community/aws/plugins/modules/aws_acm_info.py
Pipelining is enabled.
<testhost> ESTABLISH LOCAL CONNECTION FOR USER: root
<testhost> EXEC /bin/sh -c 'ANSIBLE_DEBUG_BOTOCORE_LOGS=True /usr/bin/python3.8 && sleep 0'
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_aws_acm_info_payload_25vnfkvy/ansible_aws_acm_info_payload.zip/ansible_collections/amazon/aws/plugins/module_utils/acm.py", line 111, in get_certificates
    cert_data = self.describe_certificate_with_backoff(client, certificate['CertificateArn'])
  File "/tmp/ansible_aws_acm_info_payload_25vnfkvy/ansible_aws_acm_info_payload.zip/ansible_collections/amazon/aws/plugins/module_utils/cloud.py", line 118, in _retry_wrapper
    return _retry_func(
  File "/tmp/ansible_aws_acm_info_payload_25vnfkvy/ansible_aws_acm_info_payload.zip/ansible_collections/amazon/aws/plugins/module_utils/cloud.py", line 68, in _retry_func
    return func()
  File "/tmp/ansible_aws_acm_info_payload_25vnfkvy/ansible_aws_acm_info_payload.zip/ansible_collections/amazon/aws/plugins/module_utils/acm.py", line 82, in describe_certificate_with_backoff
    return client.describe_certificate(CertificateArn=certificate_arn)['Certificate']
  File "/usr/local/lib/python3.8/dist-packages/botocore/client.py", line 357, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/local/lib/python3.8/dist-packages/botocore/client.py", line 676, in _make_api_call
    raise error_class(parsed_response, operation_name)
```

File: tests/integration/targets/aws_acm/aliases
Comment (sebastien-rosset) [inline tests/integration/targets/aws_acm/aliases:2]: @markuman I'm unclear what to do about this label.
It looks like when `unstable` is set, the tests do not execute. But isn't this a gating criteria for the PR?
On the other hand, if the module is known to have issue such as the referenced issue, how can we test?

I've also discovered a new intermittent issue: https://github.com/ansible-collections/amazon.aws/issues/622

Comment (sebastien-rosset) [reply tests/integration/targets/aws_acm/aliases:2]: ping @markuman.

**Comment by markuman** [comment]
it looks like a bug 

```
botocore.errorfactory.ResourceNotFoundException: An error occurred (ResourceNotFoundException) when calling the DescribeCertificate operation: Could not find certificate arn:aws:acm:us-east-1:966509639900:certificate/38caaf23-9f4c-4170-8424-46367ae7b485.
fatal: [testhost]: FAILED! => {
    "boto3_version": "1.16.0",
    "botocore_version": "1.19.0",
    "changed": false,
    "error": {
        "code": "ResourceNotFoundException",
        "message": "Could not find certificate arn:aws:acm:us-east-1:966509639900:certificate/38caaf23-9f4c-4170-8424-46367ae7b485."
    },
    "failed_when_result": "The conditional check 'check_del_three.certificates | length != 0' failed. The error was: error while evaluating conditional (check_del_three.certificates | length != 0): 'dict object' has no attribute 'certificates'",
 ```

File: tests/integration/targets/aws_acm/aliases
Comment (sebastien-rosset) [reply tests/integration/targets/aws_acm/aliases:2]: What should I do next? Should I disable these integration tests because there is a bug in amazon.aws?

Here is a fix for amazon.aws: https://github.com/ansible-collections/amazon.aws/pull/646, but now I'm concerned I'm going to have to wait until there is a new release of amazon.aws that includes a fix?

**Comment by sebastien-rosset** [comment]
Can you help with this error? I'm not getting any error when I run `ansible-test sanity --python=3.10 --docker -vvvv` in my workspace.

community.aws.aws_acm missing documentation (or could not parse documentation): unknown doc_fragment(s) in file /tmp/tmps4sax5_b/ansible_collections/community/aws/plugins/modules/aws_acm.py: amazon.aws.aws, amazon.aws.ec2\n'

File: tests/integration/targets/aws_acm/aliases
Comment (markuman) [reply tests/integration/targets/aws_acm/aliases:2]: AFAIK the next amazon.aws release is scheduled one week before community.aws will release its next version.   
imo, I would try to wait this two weeks if that solves the Issue with the CI.

**Comment by sebastien-rosset** [comment]
There are CI failures because of #902

File: tests/integration/targets/aws_acm/aliases
Comment (sebastien-rosset) [reply tests/integration/targets/aws_acm/aliases:2]: Ouch. This is not encouraging.  Given you are planning to release community.aws on Feb 14th and you are suggesting to wait 2 weeks, does that mean this PR wouldn't make it for 3.1?

**Comment by markuman** [comment]
> Can you help with this error? I'm not getting any error when I run `ansible-test sanity --python=3.10 --docker -vvvv` in my workspace.
> 
> community.aws.aws_acm missing documentation (or could not parse documentation): unknown doc_fragment(s) in file /tmp/tmps4sax5_b/ansible_collections/community/aws/plugins/modules/aws_acm.py: amazon.aws.aws, amazon.aws.ec2\n'

@sebastien-rosset When your clone of `community.aws` is located in `~/ansible_collections/community/aws`, you'll also need  a clone of `amazon.aws` in `~/ansible_collections/amazon/aws`

**Comment by markuman** [comment]
> There are CI failures because of #902

[No, it's still the same error](https://github.com/ansible-collections/community.aws/pull/870#issuecomment-1023917809)  
I'll try to run the integration tests against main and your branch locally tomorrow.

File: tests/integration/targets/aws_acm/aliases
Comment (markuman) [reply tests/integration/targets/aws_acm/aliases:2]: No. afaik amazon.aws is planed to release one week before community.aws.  
Means Feb 7th.  
But since I'm not a maintainer of amazon.aws, I cannot guarantee that. That's just something I've noticed on IRC/Matrix.  

Missing the release train happens more often than one would like. There will be 3.2.0 soon I think.

File: plugins/modules/aws_acm.py
Comment (markuman) [inline plugins/modules/aws_acm.py:297]: Imo the module should not exit in check mode asap.  
It might return usefull information in even in check_mode: https://github.com/ansible-collections/community.aws/blob/main/plugins/modules/aws_acm.py#L375-L387

Comment (sebastien-rosset) [reply plugins/modules/aws_acm.py:297]: The example you provide does exactly that. The only difference is instead having an if check_mode: exit, it has if not chec_mode: do X else exit.

File: plugins/modules/aws_acm.py
Comment (marknet15) [inline plugins/modules/aws_acm.py:541]: @markuman @alinabuzachis what do you think about adding a deprecation notice for the attributes mentioned here ? Maybe one for a separate PR.

Comment (sebastien-rosset) [reply plugins/modules/aws_acm.py:541]: Can this be handled separately?

Comment (alinabuzachis) [reply plugins/modules/aws_acm.py:541]: @marknet15 I guess we should add a deprecation notice. I am still confused because `signature_algorithms` doesn't seem to be documented. So, I guess we need to update the documentation with this param and the new one that replaces it and add a deprecation notice for `signature_algorithms`. Not sure if this makes sense @markuman .

Comment (sebastien-rosset) [reply plugins/modules/aws_acm.py:541]: What else do I need to do? This is stalling.

Comment (alinabuzachis) [reply plugins/modules/aws_acm.py:541]: @markuman how would you suggest to proceed here?

Comment (sebastien-rosset) [reply plugins/modules/aws_acm.py:541]: Can this conversation be resolved? I don't see any change being requested for this PR.

**Review by markuman — APPROVED** [review]

File: plugins/modules/aws_acm.py
Comment (markuman) [inline plugins/modules/aws_acm.py:176]:
```suggestion
    version_added: 3.2.0
```

File: plugins/modules/aws_acm.py
Comment (markuman) [inline plugins/modules/aws_acm.py:183]:
```suggestion
    version_added: 3.2.0
```

...if we get another review

**Review by alinabuzachis — APPROVED** [review]
I guess we only need to change` version_added: 3.3.0`. Other than that, LGTM!

**Pull request merged**
